### PR TITLE
WIP: isrealB / AA interface

### PR DESCRIPTION
**Pull request by cortner**
@MatthiasSachs @zhanglw0521  - tests are not passing yet, but I may not be able to finish debugging, so here is a first draft. 

I've kept the `isreal` flag after all for a few reasons: (1) maybe we want the option to overwrite it; (2) it means it gets a lot harder to put something other than `real` or `identity` in there; and (3) because this is also in `valtype` it gets harder to dispatch the allocations correctly. Please look at the changes and see if you spot any issues with what I'm proposing now.

### CODE REVIEW

File: src/pibasis.jl [hunk 1]
```diff
@@ -153,7 +153,7 @@ PIBasis(basis1p, Bsel::AbstractBasisSelector; kwargs...) =
    PIBasis(basis1p, O3(), Bsel; kwargs...)
 
 PIBasis(basis1p, symgrp, Bsel::AbstractBasisSelector; 
-        isreal = true, kwargs...) =
+        isreal = false, kwargs...) =
```
Comment (cortner) [inline src/pibasis.jl:156]: I've switched the default for the AA basis to complex

File: src/properties.jl [hunk 2]
```diff
@@ -60,6 +60,10 @@ end
 
 Base.show(io::IO, φ::Invariant) = print(io, "i($(φ.val))")
 
+isrealB(::Invariant{<: Real}) = true
```
Comment (cortner) [inline src/properties.jl:63]: this is applied after the symmetrisation

File: src/properties.jl [hunk 3]
```diff
@@ -60,6 +60,10 @@ end
 
 Base.show(io::IO, φ::Invariant) = print(io, "i($(φ.val))")
 
+isrealB(::Invariant{<: Real}) = true 
+isrealB(::Invariant{<: Complex}) = false 
+isrealAA(::Invariant{<: Real}) = true
```
Comment (cortner) [inline src/properties.jl:65]: this is applied after the product basis and before the symmetrisation

File: src/properties.jl [hunk 4]
```diff
@@ -144,6 +148,10 @@ real(φ::EuclideanVector) = EuclideanVector(real(φ.val))
 complex(φ::EuclideanVector) = EuclideanVector(complex(φ.val))
 complex(::Type{EuclideanVector{T}}) where {T} = EuclideanVector{complex(T)}
 
+# TODO MATTHIAS - IS THIS RIGHT, OR SHOULD IT DEPEND ON T? 
+isrealB(::EuclideanVector) = true 
+isrealAA(::EuclideanVector) = false 
+
```
Comment (cortner) [inline src/properties.jl:154]: @MatthiasSachs Please check that you agree

File: src/properties.jl [hunk 5]
```diff
@@ -210,6 +218,10 @@ end
 # # differentiation - cf #27
 # *(φ::SphericalVector, dAA::SVector) = φ.val * dAA'
 
+isrealB(::SphericalVector) = false 
+isrealAA(::SphericalVector) = false 
+
```
Comment (cortner) [inline src/properties.jl:223]: @zhanglw0521 please check that you agree

Comment (cortner) [reply src/properties.jl:223]: also also below

File: src/symmbasis.jl [hunk 6]
```diff
@@ -97,10 +93,10 @@ SymmetricBasis(φ::AbstractProperty,
                basis1p::OneParticleBasis, 
                symgrp::SymmetryGroup, 
                Bsel::AbstractBasisSelector; 
-               isreal=false, kwargs...) =
+               isreal=isrealB(φ), kwargs...) =
```
Comment (cortner) [inline src/symmbasis.jl:96]: this is the first place where the `isreal` is automagically applied

File: src/symmbasis.jl [hunk 7]
```diff
@@ -97,10 +93,10 @@ SymmetricBasis(φ::AbstractProperty,
...
       SymmetricBasis(φ, symgrp, 
                      PIBasis(basis1p, symgrp, Bsel; 
-                             kwargs..., property = φ); 
+                             isreal = isrealAA(φ), kwargs..., property = φ);
```
Comment (cortner) [inline src/symmbasis.jl:99]: and here is the second place.

**Comment by cortner** [comment]
@MatthiasSachs and @zhanglw0521 I'd be grateful if you take a quick look at those changes - I'd like to merge sometime tomorrow.

File: src/properties.jl (hunk 4 above)
Comment (MatthiasSachs) [reply src/properties.jl:154]: I would think that it should depend on the in isrealB, i.e., I think we should have

isrealB(::EuclideanVector{<: Real}) = true 
isrealB(::EuclideanVector{<: Complex}) = false 
isrealAA(::EuclideanVector) = false

I will try to test this myself tomorrow!

File: src/properties.jl (hunk 5 above)
Comment (zhanglw0521) [reply src/properties.jl:223]: Thanks for the rewriting. I see no mistake both here and below.

File: src/properties.jl (hunk 4 above)
Comment (cortner) [reply src/properties.jl:154]: Ok, but only if having complex covariant vectors makes sense at all? Does it? Or should Euclidean covariant vectors just be restricted to be real?

Comment (cortner) [reply src/properties.jl:154]: (The fewer code-paths the fewer bugs…)

Comment (cortner) [reply src/properties.jl:154]: Anyhow if you want both real and complex variants then we also need tests for both

Comment (MatthiasSachs) [reply src/properties.jl:154]: > Ok, but only if having complex covariant vectors makes sense at all? Does it? Or should Euclidean covariant vectors just be restricted to be real?

yes, I am wondered about that myself. Do we have applications for that?

Comment (cortner) [reply src/properties.jl:154]: I can’t think of any. If we want two vectors then we just get two vectors.//

Comment (MatthiasSachs) [reply src/properties.jl:154]: OK,  I think there is an issue: in the current implementation we have to initialize an euclidian vector by
φ = ACE.EuclideanVector(Complex{Float64})
irrespectively of whether in the end we want the basis functions itself to be real or complex. I think this is very confusing. I in order to resolve the I think we either need to require EuclidianVector to be always real (as in your current implementation) or we need to add an additional parameter to EuclidianVector which indicates whether the property is supposed to be treated as complex or real and on which we can dispatch. Does this make sense?

Comment (cortner) [reply src/properties.jl:154]: Let’s require that it is always real, and change the type parameter to just Float64.

Comment (cortner) [reply src/properties.jl:154]: Can tou have a go at this? Ping me if you run into problems

Comment (MatthiasSachs) [reply src/properties.jl:154]: OK, I have a working version ready on co/realsymm. Should I push directly to that branch?

Comment (cortner) [reply src/properties.jl:154]: Yes please - also remove the comment with my q for you

**Comment by cortner** [comment]
@MatthiasSachs did you push your commits into this branch? I don't see them?

**Comment by MatthiasSachs** [comment]
Sorry, just pushed the changes. I wanted to rerun the test. But somehow it took too long to load some functions. If it doesn't pass the tests here I will take another look. 
Certainly, there some question about whether or not the new implementation is more intuitive. For example, the functions  `real` and `complex` are now implemented as follows:

real(φ::EuclideanVector) = EuclideanVector(φ.val)
complex(φ::EuclideanVector) = EuclideanVector(φ.val)

File: src/properties.jl (hunk 4 above)
Comment (cortner) [reply src/properties.jl:153]: shouldn't there be other changes as well?

Comment (cortner) [reply src/properties.jl:153]: ah-sorry, I missed the previous commit

**Comment by cortner** [comment]
@MatthiasSachs Can you please configure your editor so that it doesn't add / remove arbitrary whitespace? This makes it really difficult to review the changes.

File: src/properties.jl (hunk 4 above)
Comment (cortner) [reply src/properties.jl:153]: looks all got, just running the tests now

**Comment by cortner** [comment]
I'm happy and will merge now. If bugs emerge we deal with them. The structure is more important right now.

**Comment by MatthiasSachs** [comment]
> @MatthiasSachs Can you please configure your editor so that it doesn't add / remove arbitrary whitespace? This makes it really difficult to review the changes.

Ok, I just modified the whitespace options in Atoms. I hope that will do the job.

**Pull request merged**